### PR TITLE
feat(terraform-provider): Add secondary release repo 

### LINKED ISSUE #5065: google-beta terraform provider doesn't get upgraded
**What Renovate type are you using?**

hosted, github

**Describe the bug**

Other Terraform providers get upgraded (`google` does ), but `google-beta` doesn't.

**Did you see anything helpful in debug logs?**
```
DEBUG: terraform-provider.getDependencies()
{
  "lookupName": "google"
}
DEBUG: terraform-provider.getDependencies()
{
  "lookupName": "google-beta"
}
DEBUG: terraform-provider.getDependencies()
{
  "lookupName": "archive"
}
```
...
```
DEBUG: Response code 404 (Not Found)
{
  "err": {
    "name": "HTTPError",
    "host": "registry.terraform.io",
    "hostname": "registry.terraform.io",
    "method": "GET",
    "path": "/v1/providers/hashicorp/google-beta",
    "protocol": "https:",
    "url": "https://registry.terraform.io/v1/providers/hashicorp/google-beta",
    "gotOptions": {
      "path": "/v1/providers/hashicorp/google-beta",
      "protocol": "https:",
      "slashes": true,
      "auth": null,
      "host": "registry.terraform.io",
      "port": null,
      "hostname": "registry.terraform.io",
      "hash": null,
      "search": null,
      "query": null,
      "pathname": "/v1/providers/hashicorp/google-beta",
      "href": "https://registry.terraform.io/v1/providers/hashicorp/google-beta",
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application/json",
        "accept-encoding": "gzip, deflate"
      },
      "hooks": {
        "beforeError": [],
        "init": [],
        "beforeRequest": [],
        "beforeRedirect": [],
        "beforeRetry": [],
        "afterResponse": []
      },
      "retry": {
        "methods": {},
        "statusCodes": {},
        "errorCodes": {}
      },
      "decompress": true,
      "throwHttpErrors": true,
      "followRedirect": true,
      "stream": false,
      "form": false,
      "json": true,
      "cache": false,
      "useElectronNet": false,
      "hostType": "terraform",
      "method": "GET"
    },
    "statusCode": 404,
    "statusMessage": "Not Found",
    "headers": {
      "server": "Cowboy",
      "content-encoding": "gzip",
      "content-type": "application/json",
      "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
      "via": "1.1 vegur, 1.1 varnish, 1.1 varnish",
      "accept-ranges": "bytes, bytes, bytes, bytes",
      "age": "0",
      "content-length": "49",
      "date": "Thu, 26 Dec 2019 11:10:34 GMT",
      "connection": "close",
      "x-served-by": "cache-iad2135-IAD, cache-sea4472-SEA",
      "x-cache": "MISS, MISS",
      "x-cache-hits": "0, 0",
      "vary": "Accept-Encoding"
    },
    "body": {
      "errors": [
        "Not Found"
      ]
    },
    "message": "Response code 404 (Not Found)",
    "stack": "HTTPError: Response code 404 (Not Found)\n    at EventEmitter.emitter.on (/home/ubuntu/renovateapp/node_modules/got/source/as-promise.js:74:19)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"
  }
}
INFO: Failed to look up dependency google-beta (google-beta)(packageFile="REDACTED", dependency="google-beta")
INFO: Failed to look up dependency google-beta (google-beta)(packageFile="REDACTED", dependency="google-beta")
INFO: Failed to look up dependency google-beta (google-beta)(packageFile="REDACTED", dependency="google-beta")
```

**To Reproduce**

Put this in a .tf file, filling in the project ID:
```
 provider "google-beta" {
   project = "<PROJECT_ID>"
   version = "2.20.0"
 }
```

**Additional context**

My understanding is that google publishes their beta and non-beta modules in an unusual way. It might be necessary to just hardcode some map from `google-beta` to `google` in the upgrade checking logic.

### PR DESCRIPTION
**Pull request by secustor**
This PR adds `releases.hashicorp.com` as secondary repo for builds. 

The primary source is still `registry.terraform.io` as it has more meta information e.g. source code repositories, changelogs, ...

Closes #5065

### CODE REVIEW

File: lib/datasource/terraform-provider/index.ts [hunk 1]
```diff
@@ -51,23 +73,54 @@ export async function getReleases({
...
-      });
-      return null;
+    if (
+      err.statusCode === 404 ||
+      err.code === 'ENOTFOUND' ||
+      err.stack.startsWith('TypeError')
+    ) {
+      try {
+        // registry returns an error, falling back to releases.hashicorp.com
+        const res = (
```
Comment (rarkins) [inline lib/datasource/terraform-provider/index.ts:89]: I suggest it's best to move this "catch" logic into a separate function that's called from here.

**Comment by rarkins** [comment]
Hi @secustor, can you refactor this a little before we focus on tests?

1. Move the existing registry logic out of getReleases() and into its own function, e.g. queryRegistry()
2. Use a single cache entry that covers both

e.g. the logic becomes

- if cached entry exists then return it
- look up registry
- if registry fails, look up releases
- cache result and return it

**Comment by secustor** [comment]
Rebased and updated to the new mocking setup. 

Seems like the second nock scope is breaking the other unit tests.

**Comment by JamieMagee** [comment]
@secustor I fixed a couple of mocks you were missing. With the fallback logic you implemented, we need to mock both possible requests.

**Comment by rarkins** [comment]
@secustor are there any cases where results could be found in *both* registries and we'd prefer to combine results?

**Comment by secustor** [comment]
@JamieMagee Thx for your support. 

@rarkins Not really the secondary repo contains only the available version and cpu architectures for the providers.

**Comment by rarkins** [comment]
Can you check out the recent commit I made which adds the concept registryStrategy = ‘hunt’? It simplifies the datasource implementations by pushing some logic into the index dispatcher. 

Let me know if you’re interested to refactor as part of this PR or you don’t mind me refactoring it after.

**Comment by secustor** [comment]
@rarkins I have refactored the changes to this new strategy system.

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**